base/collection_n1ql_common.go

```diff
@@ -213,51 +244,31 @@ func waitForIndexExistence(ctx context.Context, store N1QLStore, indexName strin
 
 // BuildDeferredIndexes issues a build command for any deferred sync gateway indexes associated with the bucket.
 func BuildDeferredIndexes(ctx context.Context, s N1QLStore, indexSet []string) error {
-
 	if len(indexSet) == 0 {
 		return nil
 	}
 
-	// Only build indexes that are in deferred state.  Query system:indexes to validate the provided set of indexes
-	statement := fmt.Sprintf("SELECT indexes.name, indexes.state FROM system:indexes WHERE indexes.keyspace_id = '%s'", s.IndexMetaKeyspaceID())
+	InfofCtx(ctx, KeyQuery, "Building deferred indexes: %v", indexSet)
 
-	if s.IndexMetaBucketID() != "" {
-		statement += fmt.Sprintf("AND indexes.bucket_id = '%s' ", s.IndexMetaBucketID())
-	}
-	if s.IndexMetaScopeID() != "" {
-		statement += fmt.Sprintf("AND indexes.scope_id = '%s' ", s.IndexMetaScopeID())
+	// the provided indexes can be in a state that is not yet ready to take a build command
+	// there's a delay between the time of index creation and when it's actually found in the system:indexes table
+	// this results in buildIndexes returning a not found error for an index that was very recently created
+	worker := func() (shouldRetry bool, err error, value interface{}) {
+		err = buildIndexes(ctx, s, indexSet)
+		if IsIndexNotFoundError(err) {
+			DebugfCtx(ctx, KeyQuery, "Index not found error when building indexes - will retry: %v", err)
+			return true, err, nil
+		}
+		return err != nil, err, nil
 	}
-
-	statement += fmt.Sprintf("AND indexes.name IN [%s]", StringSliceToN1QLArray(indexSet, "'"))
-	// mod: bucket name
-
-	results, err := s.executeQuery(statement)
+	sleeper := CreateDoublingSleeperDurationFunc(500, time.Second*30)
+	err, _ := RetryLoop(ctx, "BuildDeferredIndexes", worker, sleeper)
 	if err != nil {
 		return err
 	}
-	deferredIndexes := make([]string, 0)
-	var indexInfo struct {
-		Name  string `json:"name"`
-		State string `json:"state"`
-	}
-	for results.Next(ctx, &indexInfo) {
-		// If index is deferred (not built), add to set of deferred indexes
-		if indexInfo.State == IndexStateDeferred {
-			deferredIndexes = append(deferredIndexes, indexInfo.Name)
-		}
-	}
-	closeErr := results.Close()
-	if closeErr != nil {
-		return closeErr
-	}
 
-	if len(deferredIndexes) == 0 {
-		return nil
-	}
-
-	InfofCtx(ctx, KeyQuery, "Building deferred indexes: %v", deferredIndexes)
-	buildErr := buildIndexes(ctx, s, deferredIndexes)
-	return buildErr
+	InfofCtx(ctx, KeyQuery, "Built deferred indexes: %v", indexSet)
```

Comment:
The old message should be right:

```
InfofCtx(ctx, KeyQuery, "Building deferred indexes: %v", deferredIndexes)
```

This only issues a build command, but doesn't actually mean the indexes are built yet.

Reply:
Removed, since we already had the "Building" line above.